packages/gator-permissions-snap/src/core/permissionHandler.ts

```diff
@@ -196,8 +232,63 @@ export class PermissionHandler<
       updateContext: (args: { updatedContext: TContext }) => Promise<void>;
     }) => {
       let currentContext = initialContext;
-      const rerender = async () =>
+      const rerender = async () => {
         await updateContext({ updatedContext: currentContext });
+      };
+
+      // fetchAccountBalanceCallCounter is used to cancel any previously executed instances of this function.
+      let fetchAccountBalanceCallCounter = 0;
+      const fetchAccountBalance = async (context: TContext) => {
+        fetchAccountBalanceCallCounter += 1;
+        const currentFetchAccountBalanceCallCounter =
+          fetchAccountBalanceCallCounter;
+
+        const { address } = fromCaip10Address(context.accountAddressCaip10);
+
+        const { assetAddress, chainId } = fromCaip19Address(
+          context.tokenAddressCaip19,
+        );
+
+        const { balance, decimals } =
+          await this.#tokenMetadataService.getTokenBalanceAndMetadata({
+            chainId,
+            account: address,
+            assetAddress,
+          });
+
+        // only update the token balance if fetchAccountBalance hasn't been called again
+        if (
+          currentFetchAccountBalanceCallCounter ===
+          fetchAccountBalanceCallCounter
+        ) {
+          this.#tokenBalance = formatUnits({ value: balance, decimals });
+
+          // send the request to fetch the fiat balance, then re-render the UI while we wait for the response
+          const fiatBalanceRequest =
+            this.#tokenPricesService.getCryptoToFiatConversion(
+              context.tokenAddressCaip19,
+              bigIntToHex(balance),
+              context.tokenMetadata.decimals,
+            );
+
+          await rerender();
+
+          const fiatBalance = await fiatBalanceRequest;
+
+          // only update the fiat balance if fetchAccountBalance hasn't been called again
+          if (
```

Comment:
I think it would be more readable if instead of nested ifs here if you go with:
```
// The function was called again: exit without updating since we have "old" information.
if (currentFetchAccountBalanceCallCounter !== fetchAccountBalanceCallCounter) {
  return;
}
```

Also maybe change the name of variable: `currentFetchAccountBalanceCallCounter` to something like `callId` or something like that.

Reply:
yeah, good call on the return statement - I got lost in the nesting and thought I wasn't returning from the function!